File: tools/core070_namespace_2_batch.R
# Retained evidence runner for the "namespace-2" manifest-area batch (9
# EXECUTABLE_NOW cases + 4 negative controls; see
# docs/dev-log/core070/namespace-2-batch-contract.json for the full case
# list, the 47 NEEDS_NEW_JULIA_SURFACE deferrals, and the 34
# REUSED_OR_RECLASSIFY rows carried over verbatim from the manifest draft's
# own `reclassify` proposals).
#
# Design follows tools/core070_postfit_policy_batch.R's repair pattern: THIS
# R process (the one with the frozen gllvmTMB library loaded) does 100% of
# the live R-side computation itself and hands the Julia child a plain JSON
# oracle file; the Julia child (tools/core070_namespace_2_batch.jl) runs zero
# R code and carries no RCall dependency at all.
#
# Two cases (gllvm_julia_setup / gllvm_julia_fit bridge admission) are
# deliberately scoped to avoid embedding a live JuliaCall session inside this
# R process -- see the contract's `rationale_notes` for why (the same class
# of live-cross-language-embed fragility that forced the postfit-policy
# repair). Two more cases (lognormal / truncated_poisson family-bridge) are
# pure R-side admission-gate rejection checks and touch neither a live model
# fit nor the Julia child at all.
#
# argv (matches the contract's runner.outer_argv and the --self-test local
# smoke, which never reaches this file):
#   Rscript --vanilla tools/core070_namespace_2_batch.R <frozen-library> <destination>
#
# <frozen-library> is an R library directory containing an installed
# gllvmTMB built from the pinned reference commit (b4d5fee...) -- the FROZEN,
# INSTALLED library, matching masks_known.R's arg 1, not a source tree.
# <destination> must not already exist; it is created and holds
# r-oracle.json, julia-results.json, julia-stdout.log, julia-stderr.log,
# results.tsv, diagnostics.log, and receipt.json.

args <- commandArgs(TRUE)
stopifnot(length(args) == 2L)
frozen_library <- normalizePath(args[[1]], mustWork = TRUE)
output_dir <- args[[2]]
stopifnot(!dir.exists(output_dir))
dir.create(output_dir, recursive = TRUE)

sha256_file <- function(path) {
  command <- if (nzchar(Sys.which("sha256sum"))) "sha256sum" else "shasum"
  argv <- if (identical(command, "sha256sum")) path else c("-a", "256", path)
  line <- system2(command, argv, stdout = TRUE, stderr = TRUE)
  stopifnot(is.null(attr(line, "status")), length(line) >= 1L)
  sub("[[:space:]].*$", "", line[[1L]])
}

.libPaths(c(frozen_library, .libPaths()))
suppressPackageStartupMessages(library(gllvmTMB))
suppressPackageStartupMessages(library(jsonlite))
stopifnot(normalizePath(find.package("gllvmTMB")) ==
          normalizePath(file.path(frozen_library, "gllvmTMB")))

root <- normalizePath(".")
contract_path <- file.path(root, "docs/dev-log/core070/namespace-2-batch-contract.json")
contract <- jsonlite::read_json(contract_path, simplifyVector = FALSE)
contract_sha256 <- sha256_file(contract_path)

stopifnot(identical(contract$reference_commit, "b4d5fee64def88bc768dda1f1f77c29b295edd86"),
          identical(contract$status, "FROZEN_NAMESPACE_2_BATCH_CONTRACT"),
          length(contract$cases) == contract$expected_case_count,
          contract$expected_case_count == 9L,
          length(contract$negative_controls) >= 2L)

# --- validate pinned R source ------------------------------------------
source_root <- file.path(root, ".unlazy/core070-aghq/oracle-source/readback")
for (rel in names(contract$source_pins)) {
  path <- file.path(source_root, rel)
  stopifnot(file.exists(path))
  digest <- sha256_file(path)
  stopifnot(identical(digest, contract$source_pins[[rel]]))
}

# ---------------------------------------------------------------------------
# 0. Pure-R admission-gate mapper, sourced directly from the pinned
#    R/julia-bridge.R (no gllvmTMB package build required for this half --
#    it works even without the frozen library, see the batch's local smoke).
#    Used for the dispatch-key checks and the two rejection-path cases.
# ---------------------------------------------------------------------------
.ns2_bridge_env <- new.env()
sys.source(file.path(source_root, "R/julia-bridge.R"), envir = .ns2_bridge_env)
`%||%` <- function(a, b) if (is.null(a)) b else a
assign("%||%", get("%||%"), envir = .ns2_bridge_env)

.ns2_family_key <- function(family_string) {
  tryCatch(list(ok = TRUE, key = .ns2_bridge_env$.gllvm_julia_family_scalar(family_string), error = ""),
            error = function(e) list(ok = FALSE, key = NA_character_, error = conditionMessage(e)))
}

# ---------------------------------------------------------------------------
# 1. Gaussian fixture (same convention as tools/core070_postfit_policy_batch.R;
#    a fresh, independently-seeded draw).
# ---------------------------------------------------------------------------
set.seed(42)
p <- 5L; K <- 2L; n <- 80L
Lambda_true <- matrix(c(
  0.8,  0.0,
  0.5,  0.6,
  0.3, -0.4,
 -0.2,  0.5,
  0.1,  0.3
), nrow = p, ncol = K, byrow = TRUE)
sigma_true <- 0.7
eta_g <- matrix(rnorm(K * n), nrow = K, ncol = n)
Y_g <- Lambda_true %*% eta_g + sigma_true * matrix(rnorm(p * n), nrow = p, ncol = n)
Y_g <- Y_g - rowMeans(Y_g)

trait_names_g <- paste0("t", seq_len(p))
df_long_g <- data.frame(
  site  = factor(rep(seq_len(n), each = p)),
  trait = factor(rep(trait_names_g, times = n), levels = trait_names_g),
  value = as.vector(Y_g)
)

fit_g <- gllvmTMB(
  value ~ 0 + trait + latent(0 + trait | site, d = K, unique = FALSE),
  data = df_long_g, unit = "site", trait = "trait", family = gaussian(),
  control = gllvmTMBcontrol(n_init = 1L, se = FALSE)
)
stopifnot("gllvmTMB_multi" %in% class(fit_g))
oracle_g <- list(
  coef = as.numeric(coef(fit_g)),
  loglik = as.numeric(logLik(fit_g))
)

# --- gllvmTMB_wide consistency (pure R-internal, no Julia involved) --------
# Batch-spec repair (2026-09-01): gllvmTMB_wide constructs
# `latent(0 + trait | site, d = K)` at the DEFAULT `unique`, whereas oracle_g
# above is fit with `unique = FALSE` — a smaller model (15 vs 19 df on this
# fixture; measured logLik gap 6.38). Comparing the wide wrapper against the
# unique=FALSE oracle was a wrong-model comparison in the original batch spec.
# The correct reference is a long-formula fit with the same default `unique`;
# measured agreement on this fixture is ~1e-8 (two independent optimizations),
# so the check tolerance is 1e-6 — calibrated at authoring time, not a
# widening of any previously accepted contract.
fit_g_default_unique <- gllvmTMB(
  value ~ 0 + trait + latent(0 + trait | site, d = K),
  data = df_long_g, unit = "site", trait = "trait", family = gaussian(),
  control = gllvmTMBcontrol(n_init = 1L, se = FALSE)
)
Y_wide <- t(Y_g)                      # n_sites x n_species, matching gllvmTMB_wide's convention
colnames(Y_wide) <- trait_names_g
fit_wide <- gllvmTMB_wide(Y_wide, d = K, family = gaussian(),
                           control = gllvmTMBcontrol(n_init = 1L, se = FALSE))
loglik_wide <- as.numeric(logLik(fit_wide))
wide_delta <- abs(loglik_wide - as.numeric(logLik(fit_g_default_unique)))

# ---------------------------------------------------------------------------
# 2. Negative-binomial fixture (shared by nbinom1 and nbinom2 fits).
# ---------------------------------------------------------------------------
set.seed(4201)
p_nb <- 5L; K_nb <- 1L; n_nb <- 60L
Lambda_nb <- matrix(c(0.6, 0.4, -0.3, 0.5, 0.2), nrow = p_nb, ncol = K_nb)
eta_nb <- matrix(rnorm(K_nb * n_nb), nrow = K_nb, ncol = n_nb)
trait_int_nb <- c(0.5, 0.2, -0.1, 0.3, 0.0)
mu_lin <- Lambda_nb %*% eta_nb + trait_int_nb
mu_nb <- exp(mu_lin)
Y_nb <- matrix(rnbinom(p_nb * n_nb, mu = as.vector(mu_nb), size = 4), nrow = p_nb, ncol = n_nb)

trait_names_nb <- paste0("t", seq_len(p_nb))
df_long_nb <- data.frame(
  site  = factor(rep(seq_len(n_nb), each = p_nb)),
  trait = factor(rep(trait_names_nb, times = n_nb), levels = trait_names_nb),
  count = as.vector(Y_nb)
)

fit_nb1 <- gllvmTMB(
  count ~ 0 + trait + latent(0 + trait | site, d = K_nb, unique = FALSE),
  data = df_long_nb, unit = "site", trait = "trait", family = nbinom1(),
  control = gllvmTMBcontrol(n_init = 1L, se = FALSE)
)
fit_nb2 <- gllvmTMB(
  count ~ 0 + trait + latent(0 + trait | site, d = K_nb, unique = FALSE),
  data = df_long_nb, unit = "site", trait = "trait", family = nbinom2(),
  control = gllvmTMBcontrol(n_init = 1L, se = FALSE)
)
oracle_nb1 <- list(loglik = as.numeric(logLik(fit_nb1)))
oracle_nb2 <- list(loglik = as.numeric(logLik(fit_nb2)))

# ---------------------------------------------------------------------------
# 3. Ordinal (probit) fixture.
# ---------------------------------------------------------------------------
set.seed(4301)
p_ord <- 4L; K_ord <- 1L; n_ord <- 60L; n_cat <- 3L
Lambda_ord <- matrix(c(0.7, -0.5, 0.4, 0.3), nrow = p_ord, ncol = K_ord)
eta_ord <- matrix(rnorm(K_ord * n_ord), nrow = K_ord, ncol = n_ord)
lin_ord <- Lambda_ord %*% eta_ord
cuts <- c(-0.4, 0.4)
prob_to_cat <- function(x) {
  p_star <- pnorm(c(cuts, Inf) - x) - pnorm(c(-Inf, cuts) - x)
  sample.int(n_cat, size = 1, prob = pmax(p_star, 1e-8))
}
Y_ord <- matrix(vapply(as.vector(lin_ord), prob_to_cat, integer(1)), nrow = p_ord, ncol = n_ord)

trait_names_ord <- paste0("t", seq_len(p_ord))
df_long_ord <- data.frame(
  site  = factor(rep(seq_len(n_ord), each = p_ord)),
  trait = factor(rep(trait_names_ord, times = n_ord), levels = trait_names_ord),
  cat   = factor(as.vector(Y_ord), levels = seq_len(n_cat), ordered = TRUE)
)

fit_ord <- gllvmTMB(
  cat ~ 0 + trait + latent(0 + trait | site, d = K_ord, unique = FALSE),
  data = df_long_ord, unit = "site", trait = "trait", family = ordinal_probit(),
  control = gllvmTMBcontrol(n_init = 1L, se = FALSE)
)
oracle_ord <- list(loglik = as.numeric(logLik(fit_ord)))

# ---------------------------------------------------------------------------
# 4. Admission-gate dispatch-key results (pure R, no live fit).
# ---------------------------------------------------------------------------
gate_gaussian <- .ns2_family_key("gaussian")
gate_nb1      <- .ns2_family_key("nbinom1")
gate_nb2      <- .ns2_family_key("nbinom2")
gate_ordprob  <- .ns2_family_key("ordinal_probit")
gate_lognorm  <- .ns2_family_key("lognormal")
gate_truncpois <- .ns2_family_key("truncated_poisson")

# gllvm_julia_setup precondition check (no live JuliaCall invocation).
julia_bin_r <- Sys.which("julia")
setup_precondition_r <- nzchar(julia_bin_r)

# ---------------------------------------------------------------------------
# 5. Write the oracle JSON (all fixtures + R-side values) for the Julia child.
# ---------------------------------------------------------------------------
oracle_path <- file.path(output_dir, "r-oracle.json")
jsonlite::write_json(
  list(
    schema = "core070-namespace-2-r-oracle/v1",
    gaussian = list(p = p, n = n, K = K, y = as.numeric(Y_g),
                     coef = oracle_g$coef, loglik = oracle_g$loglik),
    gllvmtmb_wide = list(loglik_wide = loglik_wide, loglik_long = oracle_g$loglik,
                          delta = wide_delta),
    nb = list(p = p_nb, n = n_nb, K = K_nb, y = as.numeric(Y_nb),
              loglik_nbinom1 = oracle_nb1$loglik, loglik_nbinom2 = oracle_nb2$loglik),
    ordinal = list(p = p_ord, n = n_ord, K = K_ord, n_categories = n_cat,
                    y = as.integer(Y_ord), loglik = oracle_ord$loglik),
    gate = list(
      gaussian = gate_gaussian, nbinom1 = gate_nb1, nbinom2 = gate_nb2,
      ordinal_probit = gate_ordprob, lognormal = gate_lognorm,
      truncated_poisson = gate_truncpois
    ),
    setup_precondition_r = setup_precondition_r,
    julia_bin_r = julia_bin_r
  ),
  oracle_path, auto_unbox = TRUE, pretty = TRUE, null = "null", na = "null", digits = 17
)

# --- invoke the Julia child ---------------------------------------------
julia_out <- file.path(output_dir, "julia-results.json")
julia_env <- c(CORE070_NAMESPACE_2_R_ORACLE = oracle_path)
julia_bin <- Sys.which("julia")
stopifnot(nzchar(julia_bin))
old_env <- Sys.getenv(names(julia_env), unset = NA, names = TRUE)
do.call(Sys.setenv, as.list(julia_env))
t0 <- Sys.time()
julia_status <- tryCatch(
  system2(julia_bin, c("--project=.", "tools/core070_namespace_2_batch.jl", julia_out),
          stdout = file.path(output_dir, "julia-stdout.log"),
          stderr = file.path(output_dir, "julia-stderr.log")),
  finally = {
    unset <- names(old_env)[is.na(old_env)]
    if (length(unset)) Sys.unsetenv(unset)
    set <- old_env[!is.na(old_env)]
    if (length(set)) do.call(Sys.setenv, as.list(set))
  }
)
elapsed <- as.numeric(Sys.time() - t0, units = "secs")

# --- read back and evaluate ----------------------------------------------
julia_report <- if (file.exists(julia_out)) {
  jsonlite::read_json(julia_out, simplifyVector = FALSE)
} else {
  NULL
}

contract_case_ids <- vapply(contract$cases, `[[`, "", "case_id")
contract_neg_ids <- vapply(contract$negative_controls, `[[`, "", "control_id")

# gllvmTMB_wide + the two rejection-path cases are decided entirely R-side
# (never touch julia_report); merge them into the same PASS/FAIL vocabulary
# the Julia child uses for the other 6 cases before folding into one receipt.
r_only_cases <- list(
  "CORE070-NAMESPACE2-GLLVMTMB-WIDE-NATIVE-FIT" = list(
    pass = isTRUE(wide_delta <= 1e-6), delta = wide_delta),
  "CORE070-NAMESPACE2-LOGNORMAL-FAMILY-BRIDGE" = list(
    pass = isTRUE(!gate_lognorm$ok && grepl("GJL-GATE-FAMILY", gate_lognorm$error, fixed = TRUE) &&
                   grepl("lognormal", gate_lognorm$error, fixed = TRUE)),
    error = gate_lognorm$error),
  "CORE070-NAMESPACE2-TRUNCATED-POISSON-FAMILY-BRIDGE" = list(
    pass = isTRUE(!gate_truncpois$ok && grepl("GJL-GATE-FAMILY", gate_truncpois$error, fixed = TRUE) &&
                   grepl("truncated_poisson", gate_truncpois$error, fixed = TRUE)),
    error = gate_truncpois$error)
)

results_ok <- !is.null(julia_report) &&
  identical(julia_status, 0L) &&
  identical(julia_report$status, "PASS") &&
  isTRUE(julia_report$all_positive_pass) &&
  isTRUE(julia_report$negative_controls_behaved_as_expected) &&
  isTRUE(julia_report$all_checks) &&
  all(vapply(julia_report$cases, function(x) isTRUE(x$pass), logical(1))) &&
  all(vapply(r_only_cases, function(x) isTRUE(x$pass), logical(1)))

all_case_ids_seen <- if (!is.null(julia_report)) union(names(julia_report$cases), names(r_only_cases)) else names(r_only_cases)
missing_case_ids <- setdiff(contract_case_ids, all_case_ids_seen)
extra_case_ids <- setdiff(all_case_ids_seen, contract_case_ids)
results_ok <- results_ok && length(missing_case_ids) == 0 && length(extra_case_ids) == 0

raw_lines <- vapply(contract_case_ids, function(id) {
  if (id %in% names(r_only_cases)) {
    pass <- isTRUE(r_only_cases[[id]]$pass)
  } else if (!is.null(julia_report) && id %in% names(julia_report$cases)) {
    pass <- isTRUE(julia_report$cases[[id]]$pass)
  } else {
    pass <- FALSE
  }
  paste(id, if (pass) "PASS" else "FAIL", "positive", sep = "\t")
}, character(1))
neg_lines <- if (!is.null(julia_report)) {
  vapply(contract_neg_ids, function(id) {
    behaved <- isTRUE(julia_report$negative_controls[[id]]$behaved)
    paste(id, if (behaved) "PASS" else "FAIL", "negative_control", sep = "\t")
  }, character(1))
} else character(0)
raw_path <- file.path(output_dir, "results.tsv")
writeLines(c(raw_lines, neg_lines), raw_path)

diag_lines <- character(0)
if (!identical(julia_status, 0L)) diag_lines <- c(diag_lines, paste("julia_exit_code", julia_status))
if (is.null(julia_report)) diag_lines <- c(diag_lines, "julia-results.json was not written")
if (length(missing_case_ids)) diag_lines <- c(diag_lines, paste("missing_case_ids:", paste(missing_case_ids, collapse = ", ")))
if (length(extra_case_ids)) diag_lines <- c(diag_lines, paste("extra_case_ids:", paste(extra_case_ids, collapse = ", ")))
diag_path <- file.path(output_dir, "diagnostics.log")
writeLines(diag_lines, diag_path)

receipt <- list(
  status = if (isTRUE(results_ok)) "PASS" else "FAIL",
  scope = "CORE070_NAMESPACE_2_BATCH",
  reference_commit = contract$reference_commit,
  contract_sha256 = contract_sha256,
  source_pins = contract$source_pins,
  source_unchanged = TRUE,
  case_count = contract$expected_case_count,
  negative_control_count = length(contract$negative_controls),
  expected_case_ids = contract_case_ids,
  negative_control_case_ids = contract_neg_ids,
  julia_exit_code = julia_status,
  julia_elapsed_seconds = elapsed,
  julia_results_sha256 = if (file.exists(julia_out)) sha256_file(julia_out) else NA_character_,
  raw_sha256 = sha256_file(raw_path),
  diagnostics_sha256 = sha256_file(diag_path),
  r_version = R.version.string,
  gllvmTMB_version = as.character(utils::packageVersion("gllvmTMB")),
  frozen_library = frozen_library
)
receipt_path <- file.path(output_dir, "receipt.json")
jsonlite::write_json(receipt, receipt_path, auto_unbox = TRUE, pretty = TRUE, null = "null")

cat("CORE070_NAMESPACE_2_BATCH_", receipt$status, "\n", sep = "")
quit(status = if (isTRUE(results_ok)) 0L else 1L)

File: tools/core070_namespace_2_batch.jl
# Inner executor for the "namespace-2" manifest-area batch (9 planned
# EXECUTABLE_NOW cases; 6 of them decided here, the remaining 3
# -- gllvmTMB_wide native-fit consistency and the two lognormal/
# truncated_poisson rejection-path cases -- are decided entirely R-side by
# the paired outer runner and merged into the same receipt there).
#
# Pure-Julia consumer, mirroring tools/core070_postfit_policy_batch.jl: reads
# a JSON oracle file the paired R runner (tools/core070_namespace_2_batch.R)
# writes BEFORE invoking this script (that R process already has the frozen
# gllvmTMB library loaded and does 100% of the live R-side fitting itself),
# refits/dispatches the identical data natively via direct `using GLLVM`
# module calls only. No RCall, no parity-runner include, no R of any kind
# runs in this process.
#
# argv:
#   ARGS[1]  destination path for the JSON results file (parent dir need not
#            exist yet; this script mkpath()s it). Must not already exist.
#
# Env vars required (set by the outer R runner before invoking this script;
# never defaulted silently here so a misconfigured environment fails loudly
# rather than quietly running against stale or missing oracle data):
#   CORE070_NAMESPACE_2_R_ORACLE = <path to the R-written oracle JSON>
#
# Invocation:
#   julia --project=. tools/core070_namespace_2_batch.jl <out.json>

using GLLVM

# ---------------------------------------------------------------------------
# Minimal JSON reader/writer (no external dependency; mirrors the existing
# repo convention in tools/core070_postfit_policy_batch.jl).
# ---------------------------------------------------------------------------
function json_read(path::AbstractString)
    txt = read(path, String)
    pos = Ref(1)
    skip_ws!() = begin
        while pos[] <= lastindex(txt) && isspace(txt[pos[]])
            pos[] = nextind(txt, pos[])
        end
    end
    local parse_value
    function parse_string()
        pos[] += 1 # opening quote
        buf = IOBuffer()
        while true
            c = txt[pos[]]
            if c == '"'
                pos[] = nextind(txt, pos[])
                break
            elseif c == '\\'
                pos[] = nextind(txt, pos[])
                e = txt[pos[]]
                if e == 'n'; write(buf, '\n')
                elseif e == 't'; write(buf, '\t')
                elseif e == 'u'
                    hex = txt[nextind(txt, pos[]):nextind(txt, pos[], 4)]
                    write(buf, Char(parse(UInt32, hex; base = 16)))
                    pos[] = nextind(txt, pos[], 4)
                else write(buf, e)
                end
                pos[] = nextind(txt, pos[])
            else
                write(buf, c)
                pos[] = nextind(txt, pos[])
            end
        end
        String(take!(buf))
    end
    function parse_number()
        start = pos[]
        while pos[] <= lastindex(txt) && (isdigit(txt[pos[]]) || txt[pos[]] in ('-', '+', '.', 'e', 'E'))
            pos[] = nextind(txt, pos[])
        end
        s = txt[start:prevind(txt, pos[])]
        return occursin('.', s) || occursin('e', s) || occursin('E', s) ? parse(Float64, s) : parse(Int, s)
    end
    function parse_array()
        pos[] += 1
        out = Any[]
        skip_ws!()
        if txt[pos[]] == ']'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            push!(out, parse_value())
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == ']'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    function parse_object()
        pos[] += 1
        out = Dict{String, Any}()
        skip_ws!()
        if txt[pos[]] == '}'
            pos[] = nextind(txt, pos[])
            return out
        end
        while true
            skip_ws!()
            k = parse_string()
            skip_ws!()
            @assert txt[pos[]] == ':'
            pos[] = nextind(txt, pos[])
            skip_ws!()
            out[k] = parse_value()
            skip_ws!()
            if txt[pos[]] == ','
                pos[] = nextind(txt, pos[])
            elseif txt[pos[]] == '}'
                pos[] = nextind(txt, pos[])
                break
            end
        end
        out
    end
    parse_value = function ()
        skip_ws!()
        c = txt[pos[]]
        if c == '"'; return parse_string()
        elseif c == '['; return parse_array()
        elseif c == '{'; return parse_object()
        elseif c == 't'; pos[] += 4; return true
        elseif c == 'f'; pos[] += 5; return false
        elseif c == 'n'; pos[] += 4; return nothing
        else; return parse_number()
        end
    end
    skip_ws!()
    parse_value()
end

json_escape(s::AbstractString) = replace(s, "\\" => "\\\\", "\"" => "\\\"", "\n" => "\\n")
to_json(x::Bool) = x ? "true" : "false"
to_json(x::Nothing) = "null"
to_json(x::AbstractString) = "\"" * json_escape(x) * "\""
to_json(x::Integer) = string(x)
to_json(x::AbstractFloat) = isfinite(x) ? repr(x) : "null"
to_json(x::AbstractVector) = "[" * join(to_json.(x), ",") * "]"
to_json(x::AbstractDict) = "{" * join(("\"$(json_escape(string(k)))\":" * to_json(v) for (k, v) in x), ",") * "}"

length(ARGS) == 1 || error("usage: julia tools/core070_namespace_2_batch.jl <out.json>")
out_path = ARGS[1]
isfile(out_path) && error("destination already exists: $out_path")
mkpath(dirname(out_path))

@assert realpath(Base.pkgdir(GLLVM)) == realpath(pwd()) "must run from the GLLVM.jl package root"

oracle_path = get(ENV, "CORE070_NAMESPACE_2_R_ORACLE", "")
isempty(oracle_path) &&
    error("CORE070_NAMESPACE_2_R_ORACLE is required; refusing to silently default it")
isfile(oracle_path) ||
    error("CORE070_NAMESPACE_2_R_ORACLE points to a nonexistent file: $oracle_path")

oracle = json_read(oracle_path)

tol = Dict("loglik_delta" => 1e-6, "coef_delta" => 1e-6)
cases = Dict{String, Any}()

# ---------------------------------------------------------------------------
# Case 1: CORE070-NAMESPACE2-GLLVMTMB-NATIVE-FIT
# ---------------------------------------------------------------------------
g = oracle["gaussian"]
p_g = Int(g["p"]); n_g = Int(g["n"]); K_g = Int(g["K"])
y_g_flat = Float64.(g["y"])
length(y_g_flat) == p_g * n_g || error("gaussian y length mismatch")
Y_g = reshape(y_g_flat, p_g, n_g)
X_g = zeros(p_g, n_g, p_g)
for j in 1:p_g
    X_g[j, :, j] .= 1
end

fit_g = fit_gaussian_gllvm(Y_g; K = K_g, X = X_g)
j_coef_g = GLLVM.StatsAPI.coef(fit_g)
j_loglik_g = fit_g.logLik
r_coef_g = Float64.(g["coef"])
r_loglik_g = Float64(g["loglik"])

coef_delta = length(j_coef_g) == length(r_coef_g) ? maximum(abs.(j_coef_g .- r_coef_g)) : Inf
loglik_delta_g = abs(j_loglik_g - r_loglik_g)
cases["CORE070-NAMESPACE2-GLLVMTMB-NATIVE-FIT"] = Dict(
    "pass" => coef_delta <= tol["coef_delta"] && loglik_delta_g <= tol["loglik_delta"],
    "coef_delta" => coef_delta, "loglik_delta" => loglik_delta_g,
)

# ---------------------------------------------------------------------------
# Case 2: CORE070-NAMESPACE2-GLLVM-JULIA-SETUP-BRIDGE-ADMISSION
# ---------------------------------------------------------------------------
pkgdir_matches_pwd = realpath(Base.pkgdir(GLLVM)) == realpath(pwd())
julia_version_string = string(VERSION)
setup_precondition_r = oracle["setup_precondition_r"] === true
cases["CORE070-NAMESPACE2-GLLVM-JULIA-SETUP-BRIDGE-ADMISSION"] = Dict(
    "pass" => pkgdir_matches_pwd && setup_precondition_r && !isempty(julia_version_string),
    "pkgdir_matches_pwd" => pkgdir_matches_pwd,
    "setup_precondition_r" => setup_precondition_r,
    "julia_version" => julia_version_string,
)

# ---------------------------------------------------------------------------
# Case 3: CORE070-NAMESPACE2-GLLVM-JULIA-FIT-BRIDGE-ADMISSION
# ---------------------------------------------------------------------------
gate = oracle["gate"]
gate_gaussian_ok = gate["gaussian"]["ok"] === true && gate["gaussian"]["key"] == "gaussian"
bridge_key_gaussian = GLLVM._bridge_family_key("gaussian") == "gaussian"

br_g = GLLVM.bridge_fit(; y = Y_g, family = "gaussian", d = K_g, X = X_g)
bridge_loglik_delta = abs(br_g.loglik - r_loglik_g)
cases["CORE070-NAMESPACE2-GLLVM-JULIA-FIT-BRIDGE-ADMISSION"] = Dict(
    "pass" => gate_gaussian_ok && bridge_key_gaussian && bridge_loglik_delta <= tol["loglik_delta"],
    "gate_gaussian_ok" => gate_gaussian_ok, "bridge_key_gaussian" => bridge_key_gaussian,
    "bridge_loglik_delta" => bridge_loglik_delta,
)

# ---------------------------------------------------------------------------
# Cases 4-5: negative-binomial family bridge (nbinom1 -> "nb1", nbinom2 -> "negbinomial")
# ---------------------------------------------------------------------------
nb = oracle["nb"]
p_nb = Int(nb["p"]); n_nb = Int(nb["n"]); K_nb = Int(nb["K"])
y_nb_flat = Float64.(nb["y"])
length(y_nb_flat) == p_nb * n_nb || error("nb y length mismatch")
Y_nb = reshape(y_nb_flat, p_nb, n_nb)
X_nb = zeros(p_nb, n_nb, p_nb)
for j in 1:p_nb
    X_nb[j, :, j] .= 1
end
r_loglik_nb1 = Float64(nb["loglik_nbinom1"])
r_loglik_nb2 = Float64(nb["loglik_nbinom2"])

gate_nb1_ok = gate["nbinom1"]["ok"] === true && gate["nbinom1"]["key"] == "nb1"
bridge_key_nb1 = GLLVM._bridge_family_key("nb1") == "nb1"
br_nb1 = GLLVM.bridge_fit(; y = Y_nb, family = "nb1", d = K_nb, X = X_nb)
nb1_loglik_delta = abs(br_nb1.loglik - r_loglik_nb1)
cases["CORE070-NAMESPACE2-NBINOM1-FAMILY-BRIDGE"] = Dict(
    "pass" => gate_nb1_ok && bridge_key_nb1 && nb1_loglik_delta <= tol["loglik_delta"],
    "gate_nb1_ok" => gate_nb1_ok, "bridge_key_nb1" => bridge_key_nb1,
    "loglik_delta" => nb1_loglik_delta,
)

gate_nb2_ok = gate["nbinom2"]["ok"] === true && gate["nbinom2"]["key"] == "negbinomial"
bridge_key_nb2 = GLLVM._bridge_family_key("nbinom2") == "negbinomial"
br_nb2 = GLLVM.bridge_fit(; y = Y_nb, family = "negbinomial", d = K_nb, X = X_nb)
nb2_loglik_delta = abs(br_nb2.loglik - r_loglik_nb2)
cases["CORE070-NAMESPACE2-NBINOM2-FAMILY-BRIDGE"] = Dict(
    "pass" => gate_nb2_ok && bridge_key_nb2 && nb2_loglik_delta <= tol["loglik_delta"],
    "gate_nb2_ok" => gate_nb2_ok, "bridge_key_nb2" => bridge_key_nb2,
    "loglik_delta" => nb2_loglik_delta,
)

# ---------------------------------------------------------------------------
# Case 6: CORE070-NAMESPACE2-ORDINAL-PROBIT-FAMILY-BRIDGE
# ---------------------------------------------------------------------------
ord = oracle["ordinal"]
p_ord = Int(ord["p"]); n_ord = Int(ord["n"]); K_ord = Int(ord["K"])
y_ord_flat = Float64.(ord["y"])
length(y_ord_flat) == p_ord * n_ord || error("ordinal y length mismatch")
Y_ord = reshape(y_ord_flat, p_ord, n_ord)
r_loglik_ord = Float64(ord["loglik"])

gate_ord_ok = gate["ordinal_probit"]["ok"] === true && gate["ordinal_probit"]["key"] == "ordinal_probit"
bridge_key_ord = GLLVM._bridge_family_key("ordinal_probit") == "ordinal_probit"
br_ord = GLLVM.bridge_fit(; y = Y_ord, family = "ordinal_probit", d = K_ord)
ord_loglik_delta = abs(br_ord.loglik - r_loglik_ord)
cases["CORE070-NAMESPACE2-ORDINAL-PROBIT-FAMILY-BRIDGE"] = Dict(
    "pass" => gate_ord_ok && bridge_key_ord && ord_loglik_delta <= tol["loglik_delta"],
    "gate_ord_ok" => gate_ord_ok, "bridge_key_ord" => bridge_key_ord,
    "loglik_delta" => ord_loglik_delta,
)

# ---------------------------------------------------------------------------
# Negative controls: deliberately-wrong comparisons that MUST fail.
# ---------------------------------------------------------------------------
neg_coef_delta = length(j_coef_g) == length(r_coef_g) ?
    maximum(abs.(j_coef_g .- (r_coef_g .+ 1.0))) : Inf
neg_loglik_delta = abs(j_loglik_g - (r_loglik_g + 1.0))
neg_nb_swapped_delta = abs(r_loglik_nb1 - br_nb2.loglik)
neg_lognormal_accepted = gate["lognormal"]["ok"] === true  # must be false (rejection expected)

negative_controls = Dict(
    "NEG-COEF-SHIFTED" => Dict(
        "behaved" => neg_coef_delta > tol["coef_delta"], "delta" => neg_coef_delta),
    "NEG-LOGLIK-SHIFTED" => Dict(
        "behaved" => neg_loglik_delta > tol["loglik_delta"], "delta" => neg_loglik_delta),
    "NEG-LOGNORMAL-ACCEPTED" => Dict(
        "behaved" => !neg_lognormal_accepted, "gate_ok" => neg_lognormal_accepted),
    "NEG-NB-FAMILY-SWAPPED" => Dict(
        "behaved" => neg_nb_swapped_delta > tol["loglik_delta"], "delta" => neg_nb_swapped_delta),
)

all_positive_pass = all(v["pass"] for v in values(cases))
negatives_behaved = all(v["behaved"] for v in values(negative_controls))
overall_ok = all_positive_pass && negatives_behaved

report = Dict(
    "status" => overall_ok ? "PASS" : "FAIL",
    "area" => "namespace-2",
    "scope" => "CORE070_NAMESPACE_2_BATCH",
    "case_count" => length(cases),
    "negative_control_count" => length(negative_controls),
    "all_positive_pass" => all_positive_pass,
    "negative_controls_behaved_as_expected" => negatives_behaved,
    "all_checks" => overall_ok,
    "cases" => cases,
    "negative_controls" => negative_controls,
    "julia_version" => string(VERSION),
)
open(out_path, "w") do io
    print(io, to_json(report))
end
println("CORE070_NAMESPACE_2_BATCH_RESULT ",
        overall_ok ? "PASS" : "FAIL",
        " positive=", all_positive_pass, " negatives=", negatives_behaved)
overall_ok || exit(1)
